tests/datasets/test_streaming.py

```diff
@@ -0,0 +1,317 @@
+def test_dataloader_single_device(tmpdir: py.path.local, batch_size: int, drop_last: bool, num_workers: int,
+                                  persistent_workers: bool, shuffle: bool):
+    num_samples = 31
+    shard_size_limit = 1 << 6
+    samples, decoders = get_fake_samples_decoders(num_samples)
+    remote = str(tmpdir.mkdir("remote"))
+    local = str(tmpdir.mkdir("local"))
+    write_synthetic_streaming_dataset(dirname=remote, samples=samples, shard_size_limit=shard_size_limit)
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=shuffle, decoders=decoders, batch_size=batch_size)
+
+    # Build DataLoader
+    dataloader = DataLoader(dataset=dataset,
+                            batch_size=batch_size,
+                            num_workers=num_workers,
+                            drop_last=drop_last,
+                            persistent_workers=persistent_workers)
+
+    # Expected number of batches based on batch_size and drop_last
+    expected_num_batches = (num_samples // batch_size) if drop_last else math.ceil(num_samples / batch_size)
+    expected_num_samples = expected_num_batches * batch_size if drop_last else num_samples
+
+    # Iterate over DataLoader
+    rcvd_batches = 0
+    sample_order = []
+
+    for batch_ix, batch in enumerate(dataloader):
+        rcvd_batches += 1
+
+        # Every batch should be complete except (maybe) final one
+        if batch_ix + 1 < expected_num_batches:
+            assert len(batch["uid"]) == batch_size
+        else:
+            if drop_last:
+                assert len(batch["uid"]) == batch_size
+            else:
+                assert len(batch["uid"]) <= batch_size
+
+        for uid in batch["uid"]:
+            sample_order.append(int(uid))
+
+    # Test dataloader length
+    assert len(dataloader) == expected_num_batches
+    assert rcvd_batches == expected_num_batches
+
+    # Test that all samples arrived
+    assert len(sample_order) == expected_num_samples
+    if not drop_last:
+        assert len(set(sample_order)) == num_samples
+
+    # Iterate over the dataloader again to check shuffle behavior
+    second_sample_order = []
+    for batch_ix, batch in enumerate(dataloader):
+        for uid in batch["uid"]:
+            second_sample_order.append(int(uid))
+
+    assert len(sample_order) == len(second_sample_order)
+    if shuffle:
+        assert sample_order != second_sample_order
+    else:
+        assert sample_order == second_sample_order
+
+
+@pytest.mark.daily()
+@pytest.mark.timeout(10)
+@pytest.mark.world_size(2)
+@pytest.mark.parametrize("batch_size", [4])
+@pytest.mark.parametrize("drop_last", [False, True])
+@pytest.mark.parametrize("num_samples", [30, 31])
+@pytest.mark.parametrize("num_workers", [1, 3])
+@pytest.mark.parametrize("shuffle", [False, True])
+def test_dataloader_multi_device(tmpdir: py.path.local, batch_size: int, drop_last: bool, num_samples: int,
```

Comment:
Does this test need to be separate from the one above? Or can it be the same (with potentially some checks of `dist.get_global_rank() == 0` for the checks that should be happening only on the rank zero node)?

Reply:
I think the separation is useful because the test matrix is reduced by a lot if I know that single-device dataloader is working... same situation if/when we add smarter multi-node behavior, I'd like to add a `test_dataloader_multi_node`.

I could maybe see a refactor combining the two but it would run a lot more tests than necessary.

Reply:
OK, if it would be helpful to keep them seperate, that's fine -- but can the common code be moved to shared fixtures / helper functions?